src/__tests__/typeChecker.test.ts

```diff
@@ -0,0 +1,311 @@
+import { TypeAnnotatedNode, TypeAnnotatedFuncDecl } from '../types'
+import { typeToString } from '../utils/stringify'
+import { parseError } from '../index'
+import * as es from 'estree'
+
+// simple program to parse program and error if there are syntatical errors
+function parse(code: any, chapter = 1) {
+  const context = mockContext(chapter)
+  const program: any = __parse(code, context)
+  expect(program).not.toBeUndefined()
+  return validateAndAnnotate(program, context)
+}
+
+function topLevelTypesToString(program: TypeAnnotatedNode<es.Program>) {
+  return program.body
+    .filter(node => ['VariableDeclaration', 'FunctionDeclaration'].includes(node.type))
+    .map(
+      (
+        node: TypeAnnotatedNode<es.VariableDeclaration> | TypeAnnotatedNode<es.FunctionDeclaration>
+      ) => {
+        const id =
+          node.type === 'VariableDeclaration'
+            ? (node.declarations[0].id as es.Identifier).name
+            : node.id?.name!
+        const actualNode =
+          node.type === 'VariableDeclaration'
+            ? (node.declarations[0].init! as TypeAnnotatedNode<es.Node>)
+            : node
+        const type =
+          actualNode.typability === 'Untypable'
+            ? "Couldn't infer type"
+            : typeToString(
+                actualNode.type === 'FunctionDeclaration'
+                  ? (actualNode as TypeAnnotatedFuncDecl).functionInferredType!
+                  : actualNode.inferredType!
+              )
+        return `${id}: ${type}`
+      }
+    )
+    .join('\n')
+}
+
+describe('type checking pairs and lists', () => {
+  it('happy paths for list functions', () => {
+    const code1 = `
+      function accumulate(op, init, xs) {
+        return is_null(xs) ? init : op(head(xs), accumulate(op, init, tail(xs)));
+      }
+      const xs = pair(1, pair(2, null));
+      const y = accumulate((x,y)=>x+y,0,xs);
+      function map(f, xs) {
+        return is_null(xs) ? null : pair(f(head(xs)), map(f, tail(xs)));
+      }
+      const xs1 = map(x => x<4 ? true : false, xs);
+      const xs2 = map(x => x>4 ? true : false, xs);
+      function append(xs, ys) {
+          return is_null(xs) ? ys : pair(head(xs), append(tail(xs), ys));
+      }
+      const xs3 = append(xs1, xs2);
+      function remove(v, xs) {
+        return is_null(xs) ? null : v === head(xs) ? tail(xs) : pair(head(xs), remove(v, tail(xs)));
+      }
+      // const xs4 = remove(true, xs3);
+    `
+    // TODO redo how we test for this
+    // const program = typeCheck(parse(code1, 2))
+    // // @ts-ignore
+    // expect(program.body[2].declarations[0].init.inferredType).toEqual<Type>({
+    //   kind: 'primitive',
+    //   name: 'number'
+    // })
+    const [program, errors] = typeCheck(parse(code1, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+"accumulate: ((number, number) -> number, number, [number, List<number>]) -> number
+xs: [number, List<number>]
+y: number
+map: Couldn't infer type
```

Comment:
is there something wrong here? It says happy path but some types don't get inferred.

Reply:
Same as the other issue below. Resolved by reordering the statements.